### PR TITLE
4 run the tests in gh actions

### PR DESCRIPTION
**Pull request by rosepearson**
PR for using git actions to install Python dependencies, run tests and lint with a variety of Python versions.

Requirements prior to landing:
1. ~~At least one test passing~~
2. ~~Run all tests~~
3. ~~Create a strict and less strict test for windows and Linux~~
4. ~~Tidy up workflow file~~
5. ~~Ensure caching is working correctly~~

### CODE REVIEW

File: setup.py
Comment (kinow) [inline setup.py:53]: These versions can be relaxed a bit I think, e.g. was `'numpy>=1.21'`, now it could be `'numpy==1.21.*'` so that users automatically get the latest bug fixes...

**Comment by rosepearson** [comment]
Yes – definitely. This commit is only for exploring why I was getting errors when using pip git install in a conda environment – as this is the first package I’ve used this in and I am/was surprised about the install errors I was getting. I just thought I would try nail things exactly to a conda environment where the packages were installed successfully together. Anyway – I think there will be a few more spurious commits to come – I’ll move the PR from draft once I’ve moved beyond the exploration stage and addressed the issue properly 😊

**Comment by kinow** [comment]
Still lurking here! Just wanted to point out that I think you are having the exact same error I am. Your version of Ubuntu on GH actions is the same as mine (go to the latest job, then the setup step, expand the operating system section: `20.04.2`).

The "latest" Ubuntu in GH actions means the latest LTS: https://github.com/actions/virtual-environments

For Docker images, it's also good practice to point latest to a stable release, and not the latest-latest, which can be misleading. You should be able to use the latest-latest via Docker I think.

**Comment by kinow** [comment]
And here's what I used to check the `pdal` library available for each Ubuntu: https://launchpad.net/ubuntu/+source/pdal, mine focal has only `2.0.1+ds-1build3`.

I think the required C++ lib version for Python `pdal-2.2.0`, for instance, is defined [here](https://github.com/PDAL/PDAL/blob/a3dceb4eb4fd2429fdcebf3de5fda83875f48bed/CMakeLists.txt#L8) (I think?).

**Review by rosepearson — COMMENTED** [review]
> Still lurking here! Just wanted to point out that I think you are having the exact same error I am. Your version of Ubuntu on GH actions is the same as mine (go to the latest job, then the setup step, expand the operating system section: `20.04.2`).
> 
> The "latest" Ubuntu in GH actions means the latest LTS: https://github.com/actions/virtual-environments
> 
> For Docker images, it's also good practice to point latest to a stable release, and not the latest-latest, which can be misleading. You should be able to use the latest-latest via Docker I think.

Thanks for the comment! I sort of hoped you might take interest in my failing. Cool - I'd wondered if that was the case RE my error matching yours. I'm familiar with the principle of Docker, but less familiar with setting them up in practice. 

Would you recommend I try a different virtual-environment? I could try windows for instance - that is what I have on my local machine. When it comes to pinning to stable release, I guess you just google the package and take the latest stable that is supported by the OS? Thanks!

**Comment by kinow** [comment]
>Would you recommend I try a different virtual-environment?

Let me see if I can find an example with Docker + GH Actions. We have one in Cylc Flow that runs the tests against 4 different versions of `bash`, but I will look for one for Python.

>I could try windows for instance - that is what I have on my local machine.

I think it's important to test -at least- in the environment where most of your users will be using. If most of the users of GeoFabrics will be using Windows, then I think it'd make sense to focus on using GH Actions + Windows (I never tried that, so I'd find it interest!). But if you are not sure, or if most users will be on Linux, then maybe we should try to finish the current GH Actions + ubuntu pipeline.

>When it comes to pinning to stable release, I guess you just google the package and take the latest stable that is supported by the OS? Thanks!

It depends I think. But normally I guess you want the most up to date environment possible. Given that that version is compatible with the other dependencies in the operating system. Sometimes you have to wait for Ubuntu, Red Hat, or even Windows, to release patches with updates for related packages before you can update lower level tools and libraries like `gcc`, `gdal`, `pdal`, etc.

File: .github/workflows/python-test-package.yml
Comment (kinow) [inline .github/workflows/python-test-package.yml:9]: Feel free to revert this to what you had before @rosepearson , so that we can trigger it for your PR now.

File: .github/workflows/python-test-package.yml
Comment (kinow) [inline .github/workflows/python-test-package.yml:6]: I think we need `pull_request` too? So that the pull requests are tested as well.

File: .github/workflows/python-test-package.yml
Comment (kinow) [inline .github/workflows/python-test-package.yml:58]: Missing newline, not too important.

File: setup.py
Comment (kinow) [inline setup.py:53]: Now that GH actions is using Conda, these changes and the ones in `requirements.txt` can be kept or reverted. If they are fixing/improving, then no harm in leaving them here. This could be useful if GeoFabrics is released to PYPI in the future.

**Review by kinow — APPROVED** [review]
Tests passing, CI running in a reasonable time, it's also printing info about the system. I think commits can be squashed if you'd like @rosepearson , but other than that this PR looks good to me! :tada: well done fixing the CI pipeline :1st_place_medal:

File: .github/workflows/python-test-package.yml
Comment (LukeParky) [reply .github/workflows/python-test-package.yml:6]: You could add a status check to see branch is up to date before allowing merge. This would mean each PR is tested through the push workflow

File: .github/workflows/python-test-package.yml
Comment (LukeParky) [inline .github/workflows/python-test-package.yml:41]: remove unused code
```suggestion
```

File: setup.py
Comment (LukeParky) [inline setup.py:45]:
```suggestion
                      'pdal>=2.2',
```

**Review by LukeParky — CHANGES_REQUESTED** [review]
Great that you got this working, I might pick your brain about it next time you're in at the GRI.
Depending on your pull request strategy you could enforce branches being up to date for a pull request.
Also, these commits would be the perfect candidate for squashing.

Good to go after some tiny suggested changes

**Comment by rosepearson** [comment]
Awesome - Thanks @LukeParky and @kinow. I've actually still got to get the last two tests off the ground - so issues manifesting as file path errors but possibly related to something in the processor/rioxarray side of the dependency tree. I'll flick a message here once I've dealt with this and addressed your suggestions.

**Comment by kinow** [comment]
@rosepearson 

>I've actually still got to get the last two tests off the ground - so issues manifesting as file path errors but possibly related to something in the processor/rioxarray side of the dependency tree.

I had errors yesterday when running the tests in the Docker image, and fixed in this commit: https://github.com/kinow/GeoFabrics/commit/e6425586b3ae08e1fa022a6efe6ef50123327843

There's a JSON file with backward slashes for path separators, which works on Windows and I thought would work with Python's `pathlib`, but for some strange reason created a file instead of a directory (a single file with the whole name `part\part\part\etc`).

Just in case you are seeing the same error...

**Comment by rosepearson** [comment]
The test_processor_remote_tiles test is failing as the pdal writers.gdal is creating an offset dem - it overrides the exact grid location you specify with one that is nearby. This means the dem is not the same. I will see what happens if I use the same version of gdal.

**Comment by rosepearson** [comment]
> The test_processor_remote_tiles test is failing as the pdal writers.gdal is creating an offset dem - it overrides the exact grid location you specify with one that is nearby. This means the dem is not the same. I will see what happens if I use the same version of gdal.

I've been trying to make the two environments more similar to achieve more consistent behaviour. This is challenging as python-pdal is not pinned to pdal leading to[ install errors when using pdal=2.2](https://github.com/PDAL/python/issues/84).

I tied pinning all the top level dependencies but got the [install errors when using pdal=2.2](https://github.com/PDAL/python/issues/84).

Then moved to unpin everything and see if we got get tests 1&2 executing even if test 2 had differences in the resulting image.

Getting closer to consistent behaviour - pinned pdal and python pdal. Most DEM values now identical but there are still some that differ. 
image

Planned next steps?

**Comment by kinow** [comment]
I think options are

1. leave PR open until dependencies are sorted out
2. make the test less strict (probably not OK?)
3. mark the tests failing as skipped/ignored and add a blocker issue to fix them before next release
4. merge PR to have the CI running for each commit and ignore the test failures (dangerous?)

**Comment by rosepearson** [comment]
> I think options are
> 
> 1. leave PR open until dependencies are sorted out
> 2. make the test less strict (probably not OK?)
> 3. mark the tests failing as skipped/ignored and add a blocker issue to fix them before next release
> 4. merge PR to have the CI running for each commit and ignore the test failures (dangerous?)
> 
> But no idea which one would be the best option for GeoFabrics.

I'm leaning to option 1 or 3 as one of the test's is passing there is still some value in 3. I intend to spend another 1- 2 hrs trying to sorting out dependencies then I will switch to ignoring the failing tests.

File: tests/test_processor_remote_tiles/test_processor_remote_tiles.py
Comment (kinow) [inline tests/test_processor_remote_tiles/test_processor_remote_tiles.py:143]: Is it for disabling the test? If so, it's also possible to use `pytest` decorators for that: https://docs.pytest.org/en/latest/how-to/skipping.html#skipping-test-functions

When the tests are run, the skipped ones will be marked as skipped, instead of passed as it'd be here I think.

Comment (rosepearson) [reply tests/test_processor_remote_tiles/test_processor_remote_tiles.py:143]: Just for now while trying to output the artifacts so I can compare the produced DEMs and debug what is causing the differences - i.e. looks like it might be the tile ordering atm - I will remove before submitting. 

Thanks for the details on how to ignore!

File: .github/workflows/python-test-package.yml
Comment (rosepearson) [reply .github/workflows/python-test-package.yml]: good suggestion!

File: tests/test_processor_remote_tiles/test_processor_remote_tiles.py
Comment (rosepearson) [reply tests/test_processor_remote_tiles/test_processor_remote_tiles.py]: Ok - I've removed it and create a less strict linux test and a stricter windows test.

File: README.md
Comment (LukeParky) [inline README.md:11]: This is really good, do you think this could even be its own package one day? That would have pros of being more modular and would allow people to use it with fewer dependencies. Although it might just make things more complex for you for now. Something to think about post-release maybe

File: README.md
Comment (LukeParky) [inline README.md:17]: I like the use of this but it might pay to add a comment or use `$BASH_VARIABLES`, whichever you think is the best way to prevent users from thinking the command is literally `conda env create -f environment_[windows|linux].yml` instead of `conda env create -f environment_windows.yml`

File: README.md
Comment (LukeParky) [inline README.md:16]:
```suggestion
```bash
```

File: README.md
Comment (LukeParky) [inline README.md:46]: I like this documentation. Depending on the end goal at some point you may have to look at something bigger than READMEs, perhaps a wiki or some site dedicated to documentation. I don't think this is needed yet though

File: README.md
Comment (LukeParky) [inline README.md:72]: These instructions are IDE specific so it would be good to remind the reader which IDE this is for.
Or you could write it IDE-agnostically. Whichever one you prefer

Spelling here:
```suggestion
Go to 'Run>Configuration per file...' and check the **The following directory** under **Working directory settings** and specify the root of the repository with no quotes. Do this for each Python test file (usually one per test folder)
```

File: README.md
Comment (LukeParky) [inline README.md:92]:
```suggestion
[GNU GPL](https://github.com/rosepearson/GeoFabrics/LICENSE)
```

File: src/geofabrics/processor.py
Comment (LukeParky) [inline src/geofabrics/processor.py:82]: Good idea to sort these but it could be even further simplified since `sorted` will create a `list` out of any `Sequence`

```suggestion
            lidar_file_paths = sorted(pathlib.Path(self.lidar_fetcher.cache_path /
                                      self.lidar_fetcher.dataset_prefixes[lidar_dataset_index]).glob('*.laz'))
```

Comment (LukeParky) [reply src/geofabrics/processor.py:82]: If you feel this reduces readability then keep it as is

Comment (LukeParky) [reply src/geofabrics/processor.py:82]: `.glob()` will produce a list so you could also do this:

```suggestion
            lidar_file_paths = pathlib.Path(self.lidar_fetcher.cache_path /
                                      self.lidar_fetcher.dataset_prefixes[lidar_dataset_index]).glob('*.laz')
           lidar_file_paths.sort()                             
```

File: tests/test_processor_remote_tiles/test_processor_remote_tiles.py
Comment (LukeParky) [inline tests/test_processor_remote_tiles/test_processor_remote_tiles.py:130]: Good to see pytest has this functionality, good use of annotations

File: src/geofabrics/__init__.py
Comment (LukeParky) [inline src/geofabrics/__init__.py:7]: Is there a check you can add to the GitHub Actions to bump versions or perform the check. I would never remember to change this

**Review by LukeParky — CHANGES_REQUESTED** [review]
Nice one @rosepearson, I can see you're putting a lot of care into the maintainability of this project. I added a couple suggestions so feel free to start a discussion on any of them

**Review by rosepearson — COMMENTED** [review]
Thanks for all the commits and comments Luke! Some nice simplification of code in there (i.e. use sort directly).

I think I will leave the version bump manual for now - but I'm sure your right that GH actions can be used to bump it automatically tied to the issue/PR milestone I will leave that for a later time :)

Your changes to the readme are great. I'll add either a comment or $BASHVAR to the readme in regards to the environment_xx.yml.

It's a good point about the use of a wiki. The spyder specific documentation for instance would sit nicely as it's own spyder specific page. It's currently nested under a spyder heading.

File: src/geofabrics/__init__.py
Comment (rosepearson) [reply src/geofabrics/__init__.py:7]: I think I will leave it for now - but your suggestion of using GH actions is great. We could tie it to the associated PR/branch milestone. I'll look into this before we go to release

File: README.md
Comment (rosepearson) [reply README.md:46]: Great suggestion. Yes - it's definitely getting a bit long! Probably a good documentation issue.

**Comment by LukeParky** [comment]
> It's a good point about the use of a wiki. The spyder specific documentation for instance would sit nicely as it's own spyder specific page. It's currently nested under a spyder heading.

Ahh of course, I remember the heading but I couldn't scroll high enough in the GitHub diff

File: README.md
Comment (rosepearson) [reply README.md:17]: Good suggestion I'll have a quick go at clarifying it!

Comment (rosepearson) [reply README.md:17]: ok I've had a stab at making it clearer in the text above. Let me know what you think.

Comment (LukeParky) [reply README.md:17]: Looks great, just one quick suggestion then ready to go

File: README.md
Comment (LukeParky) [inline README.md:14]:
```suggestion
You can use this package by using pip to install the package and dependencies using the following commands in a conda terminal to add it to that environment, where you must either specify `environment_windows.yml` or `environment_linux.yml` depending on your operating system. Each file is located in the root repository folder. Sorry there is no macOS support at this stage.
```

**Review by LukeParky — CHANGES_REQUESTED** [review]
Just one formatting suggestion in the README and then it's ready to go. Great PR

File: README.md
Comment (rosepearson) [reply README.md:14]: Great suggestion!

**Review by LukeParky — APPROVED** [review]

**Review by LukeParky — APPROVED** [review]
Good catch

**Pull request merged**
